File: truss-compiler/src/main/java/truss/compiler/binding/TopLevelSymbolBinder.java
package truss.compiler.binding;

import org.apache.commons.lang.Validate;
import truss.compiler.symbols.ContainerSymbol;
import truss.compiler.symbols.NamespaceSymbol;
import truss.compiler.symbols.TypeSymbol;
import truss.compiler.syntax.*;

import java.util.Stack;

class TopLevelSymbolBinder extends SyntaxTreeWalker {
    private final SymbolManager manager;
    private final Stack<ContainerSymbol> stack = new Stack<>();

    public TopLevelSymbolBinder(SymbolManager manager) {
        Validate.notNull(manager, "manager");

        this.manager = manager;
        stack.push(manager.getGlobalSymbol());
    }

    @Override
    public void visitDelegateDeclaration(DelegateDeclarationSyntax syntax) throws Exception {
        manager.put(
            syntax,
            TypeSymbol.fromDelegate(syntax, stack.peek())
        );
    }

    @Override
    public void visitEnumDeclaration(EnumDeclarationSyntax syntax) throws Exception {
        manager.put(
            syntax,
            TypeSymbol.fromEnum(syntax, stack.peek())
        );
    }

    @Override
    public void visitNamespaceDeclaration(NamespaceDeclarationSyntax syntax) throws Exception {
        ContainerSymbol symbol = NamespaceSymbol.fromDeclaration(
            syntax,
            stack.peek()
        );

        manager.put(syntax, symbol);
        stack.push(symbol);

        super.visitNamespaceDeclaration(syntax);

        stack.pop();
    }

    @Override
    public void visitTypeDeclaration(TypeDeclarationSyntax syntax) throws Exception {
        ContainerSymbol symbol = TypeSymbol.fromType(
            syntax,
            stack.peek()
        );

        manager.put(syntax, symbol);
        stack.push(symbol);

        super.visitTypeDeclaration(syntax);

        stack.pop();
    }
}

File: Truss.Compiler/Binding/TopLevelSymbolBinder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Truss.Compiler.Symbols;
using Truss.Compiler.Syntax;

namespace Truss.Compiler.Binding {
    internal class TopLevelSymbolBinder : SyntaxTreeWalker {
        private readonly ErrorList _errors;
        private readonly SymbolManager _manager;
        private readonly Stack<ContainerSymbol> _stack = new Stack<ContainerSymbol>();

        public TopLevelSymbolBinder(ErrorList errors, SymbolManager manager) {
            if (errors == null) {
                throw new ArgumentNullException("errors");
            }
            if (manager == null) {
                throw new ArgumentNullException("manager");
            }

            _errors = errors;
            _manager = manager;
            _stack.Push(manager.GlobalSymbol);
        }

        public override void VisitDelegateDeclaration(DelegateDeclarationSyntax syntax) {
            _manager.Add(
                syntax,
                TypeSymbol.FromDelegate(_errors, syntax, _stack.Peek())
            );
        }

        public override void VisitEnumDeclaration(EnumDeclarationSyntax syntax) {
            _manager.Add(
                syntax,
                TypeSymbol.FromEnum(_errors, syntax, _stack.Peek())
            );
        }

        public override void VisitNamespaceDeclaration(NamespaceDeclarationSyntax syntax) {
            var symbol = NamespaceSymbol.FromDeclaration(_errors, syntax, _stack.Peek());

            _manager.Add(syntax, symbol);
            _stack.Push(symbol);

            base.VisitNamespaceDeclaration(syntax);

            _stack.Pop();
        }

        public override void VisitTypeDeclaration(TypeDeclarationSyntax syntax) {
            var symbol = TypeSymbol.FromType(_errors, syntax, _stack.Peek());

            _manager.Add(syntax, symbol);
            _stack.Push(symbol);

            base.VisitTypeDeclaration(syntax);

            _stack.Pop();
        }
    }
}
